File: Lecture2/SineNode/src/SineNode.cpp
#include "SineNode.h"
#include <maya/MFnNumericAttribute.h>
#include <maya/MFnUnitAttribute.h>
#include <maya/MFnEnumAttribute.h>
#include <maya/MTime.h>


//----------------------------------------------------------------------------------------------------------------------
/// @brief simple macro to check status and return if error
/// originally written by Sola Aina
//----------------------------------------------------------------------------------------------------------------------
#define CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( stat , message )				\
	if( !status )										\
	{											\
		MString errorString = status.errorString() + " -- " + MString( message );	\
		MGlobal::displayError( errorString );						\
		return MStatus::kFailure;							\
	}											\
//----------------------------------------------------------------------------------------------------------------------
/// @brief simple macro to check status and return if error
/// originally written by Sola Aina
//----------------------------------------------------------------------------------------------------------------------

#define CHECK_STATUS_AND_RETURN_IF_FAIL( stat , message )					\
	if( !status )										\
	{											\
		MString errorString = status.errorString() + " -- " + MString( message );	\
		MGlobal::displayError( errorString );						\
	}											\

/// @brief macro to get rid of compiler warnings
#define UNUSED(arg) (void)arg;

//----------------------------------------------------------------------------------------------------------------------
// as these are static initialisers we need to set them here.
MTypeId SineNode::m_id( 0x70001 );		// numeric Id of node
const MString SineNode::m_typeName( "sineNode" );
MObject SineNode::m_amplitude;
MObject SineNode::m_frequency;
MObject SineNode::m_time;
MObject SineNode::m_output;
MObject SineNode::m_functionType;

//----------------------------------------------------------------------------------------------------------------------
MStatus	SineNode::initialize()
{
  MStatus status;
  // the input attibute is a time attribute that will be connected to the global time node attribute,
  // time1.outTime
  //
  //	Any attribute types that are measured in changeable units, ie distance(cm, mm, miles, feet),
  //	angles( degrees, radians ), time( secs, mins, frames ) etc should be created using the
  //	MFnUnitAttribute function set.
  //
  MFnUnitAttribute uAttr;
  m_time =  uAttr.create( "time", "t", MFnUnitAttribute::kTime );
  // since the input attribute will be connected to the time value, there is little point
  // in storing this attribute within the maya binary file. There is also no point in
  // making the attribute keyable.
  uAttr.setStorable(false);
  uAttr.setKeyable(false);
  status = addAttribute( m_time );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to add \"time\" attribute to SineNode" );


  MFnEnumAttribute    enumAttr;
  // note that f is used later as the shortcut for frequency so we use ft
  m_functionType = enumAttr.create( "function", "ft", 0, &status );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to create \"function\" attribute" );

  enumAttr.addField( "sine", SINE );
  enumAttr.addField( "cos", COS );
  enumAttr.addField( "complex", COMPLEX );
  enumAttr.setHidden( false );
  enumAttr.setKeyable( true );
  status = addAttribute( m_functionType );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to add \"function type\" attribute to SineNode" );


  MFnNumericAttribute	numAttr;

  m_amplitude = numAttr.create( "amplitude", "a", MFnNumericData::kDouble, 1.0, &status );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to create \"amplitude\" attribute" );
	numAttr.setChannelBox( true );			// length attribute appears in channel box
  // add attribute
  status = addAttribute( m_amplitude );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to add \"amplitude\" attribute to SineNode" );

  // now the freq one
  m_frequency = numAttr.create( "frequency", "f", MFnNumericData::kDouble, 1.0, &status );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to create \"frequency\" attribute" );
  numAttr.setChannelBox( true );			// freq attribute appears in channel box
  // Add  attribute to node
  status = addAttribute( m_frequency );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to add \"freq\" attribute to SineNode" );



  // create the output attribute
  m_output = numAttr.create( "output", "o",MFnNumericData::kDouble,0.0f);

  // make this attribute a read only output value
  numAttr.setReadable(true);
  numAttr.setKeyable(false);
  numAttr.setWritable(false);
  status = addAttribute( m_output );
  CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to add \"output\" attribute to SineNode" );

  attributeAffects(m_time,  m_output);
  attributeAffects(m_functionType,m_output);

	return MStatus::kSuccess;
}

//----------------------------------------------------------------------------------------------------------------------
// This method should be overridden in user defined nodes.
// Recompute the given output based on the nodes inputs.
// The plug represents the data value that needs to be recomputed, and the data block holds the storage
// for all of the node's attributes.
//----------------------------------------------------------------------------------------------------------------------
MStatus SineNode::compute( const MPlug &_plug , MDataBlock &_data )
{
  MStatus status;
  if( _plug == m_output)
	{
		// get data handle for length
    MDataHandle timeData = _data.inputValue( m_time , &status );
    CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to get data handle for time plug" );

    // get data handle for function
    MDataHandle funcData = _data.inputValue( m_functionType , &status );
    CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to get data handle for function plug" );



    MDataHandle outputData = _data.outputValue(m_output);
    CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to get data handle for output plug" );

    // get the input value as a time value
    MTime Time = timeData.asTime();

    // get data handle for length
    MDataHandle dataHandle = _data.inputValue( m_frequency , &status );
    CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to get data handle for freq plug" );
    // now get the value for the data handle as a double
    double freq = dataHandle.asDouble();
    dataHandle = _data.inputValue( m_amplitude , &status );
    CHECK_STATUS_AND_RETURN_MSTATUS_IF_FAIL( status , "Unable to get data handle for amplitude plug" );
    // now get the value for the data handle as a double
    double amp = dataHandle.asDouble();
    double out;
    switch(funcData.asInt())
    {
      case SINE :  out=amp*sin(freq*M_PI*Time.as(MTime::kSeconds)); break;
      case COS :   out=amp*cos(freq*M_PI*Time.as(MTime::kSeconds)); break;
      case COMPLEX :
                out=amp*cos(freq*M_PI*Time.as(MTime::kSeconds)) +
                    amp*sin(2*freq*M_PI*Time.as(MTime::kSeconds));
      break;
    }


    // set the output to the sine of the time (in seconds)
    outputData.set(out);
    // clean the output plug, ie unset it from dirty so that maya does not re-evaluate it
    _data.setClean( _plug );

    return MStatus::kSuccess;
	}

  return MStatus::kUnknownParameter;
}

//----------------------------------------------------------------------------------------------------------------------
void* SineNode::creator()
{
  return new SineNode();
}
//----------------------------------------------------------------------------------------------------------------------

File: PythonPlugins/SineNode.py
import maya.api.OpenMaya as OpenMaya
import sys
import math

# Set this flag to show we are using api 2.0
maya_useNewAPI = True


class SineNode(OpenMaya.MPxNode):
    # class attributes are placed here maya will use these to create the node
    # and to determine the type of node which is setup during initialize method
    # Older plugins used to set these to OpenMaya.MObject() (see api1 demos)
    # however we now just set to None and use the FunctionSets to create the attributess
    id = OpenMaya.MTypeId(0x00105481)
    amplitude = None
    frequency = None
    time = None
    output = None
    function_type = None

    def __init__(self):
        OpenMaya.MPxNode.__init__(self)

    # factory to create the node
    @staticmethod
    def creator():
        return SineNode()

    @staticmethod
    def initialize():
        # Create the attributes
        numeric_attrib_fn = OpenMaya.MFnNumericAttribute()
        SineNode.amplitude = numeric_attrib_fn.create(
            "amplitude", "a", OpenMaya.MFnNumericData.kDouble, 1.0
        )
        numeric_attrib_fn.storable = True
        numeric_attrib_fn.keyable = True
        numeric_attrib_fn.readable = True
        numeric_attrib_fn.writable = True
        OpenMaya.MPxNode.addAttribute(SineNode.amplitude)
        # add frequency attribute
        SineNode.frequency = numeric_attrib_fn.create(
            "frequency", "f", OpenMaya.MFnNumericData.kDouble, 1.0
        )
        numeric_attrib_fn.storable = True
        numeric_attrib_fn.keyable = True
        numeric_attrib_fn.readable = True
        numeric_attrib_fn.writable = True
        OpenMaya.MPxNode.addAttribute(SineNode.frequency)
        # add enum attribute
        enum_attribute_fn = OpenMaya.MFnEnumAttribute()
        SineNode.function_type = enum_attribute_fn.create("functionType", "ft")
        enum_attribute_fn.addField("sine", 0)
        enum_attribute_fn.addField("cosine", 1)
        enum_attribute_fn.addField("complex", 2)

        enum_attribute_fn.storable = True
        enum_attribute_fn.keyable = True
        enum_attribute_fn.readable = True
        OpenMaya.MPxNode.addAttribute(SineNode.function_type)

        # Time attribute
        time_attribute_fn = OpenMaya.MFnUnitAttribute()
        SineNode.time = time_attribute_fn.create(
            "time", "t", OpenMaya.MFnUnitAttribute.kTime, 0.0
        )
        time_attribute_fn.storable = False
        time_attribute_fn.keyable = False
        OpenMaya.MPxNode.addAttribute(SineNode.time)
        # output set by the compute
        SineNode.output = numeric_attrib_fn.create(
            "output", "o", OpenMaya.MFnNumericData.kDouble, 0.0
        )
        numeric_attrib_fn.storable = False
        numeric_attrib_fn.keyable = False
        numeric_attrib_fn.readable = True
        numeric_attrib_fn.writable = False
        OpenMaya.MPxNode.addAttribute(SineNode.output)
        # Set the attribute dependencies
        OpenMaya.MPxNode.attributeAffects(SineNode.amplitude, SineNode.output)
        OpenMaya.MPxNode.attributeAffects(SineNode.frequency, SineNode.output)
        OpenMaya.MPxNode.attributeAffects(SineNode.time, SineNode.output)
        OpenMaya.MPxNode.attributeAffects(SineNode.function_type, SineNode.output)
        return True

    def compute(self, plug, data):
        """
        Description

           When input attributes are dirty this method will be called to
           recompute the output attributes.

        Arguments

           plug      - the attribute that triggered the compute
           data - the nodes data
        """
        # we only need to compute if the plug is the output node changing
        if plug == SineNode.output:
            # get the input data and convert to python types
            time_data = data.inputValue(SineNode.time)
            time = time_data.asTime()
            amplitude = data.inputValue(SineNode.amplitude).asDouble()
            frequency = data.inputValue(SineNode.frequency).asDouble()
            function_type = data.inputValue(SineNode.function_type).asShort()
            # compute the result
            if function_type == 0:
                result = amplitude * math.sin(
                    math.radians(frequency * math.pi * time.asUnits(OpenMaya.MTime.kSeconds))
                )
            elif function_type == 1:
                result = amplitude * math.cos(
                    math.radians(frequency * math.pi * time.asUnits(OpenMaya.MTime.kSeconds))
                )
            else:
                result = amplitude * math.sin(
                    math.radians(frequency * math.pi * time.asUnits(OpenMaya.MTime.kSeconds))
                ) + amplitude * math.sin(
                    math.radians(
                        2 * frequency * math.pi * time.asUnits(OpenMaya.MTime.kSeconds)
                    )
                )
            # set the output
            output_data = data.outputValue(SineNode.output)
            output_data.setDouble(result)
            data.setClean(plug)
            return True
        return False

    def postConstructor(self):
        """
        Description

           When instances of this node are created internally, the MObject associated
           with the instance is not created until after the constructor of this class
           is called. This means that no member functions of Node can
           be called in the constructor.
           The postConstructor solves this problem. Maya will call this function
           after the internal object has been created.
           As a general rule do all of your initialization in the postConstructor.
           Not used here but for info
        """
        print("postConstructor called")


def initializePlugin(obj):
    plugin = OpenMaya.MFnPlugin(obj)
    try:
        plugin.registerNode(
            "SineNodePy", SineNode.id, SineNode.creator, SineNode.initialize
        )
    except:
        sys.stderr.write("Failed to register node\n")
        raise


def uninitializePlugin(obj):
    plugin = OpenMaya.MFnPlugin(obj)

    try:
        plugin.deregisterNode(SineNode.id)
    except:
        sys.stderr.write("Failed to deregister node\n")
        raise
